packages/platforms/react-native/tests/retry-queue/file-based.test.ts

```diff
@@ -0,0 +1,350 @@
+      expect(delivery.requests).toHaveLength(1)
+
+      // all the files should have been deleted
+      expect(await directory.files()).toStrictEqual([])
+    })
+
+    it('deletes files with invalid contents', async () => {
+      const validEndTime = BigInt(Date.now()) * BigInt(1_000_000)
+      const payload = createPayload({ spanId: 'a', endTimeUnixNano: validEndTime.toString() })
+
+      const fileSystem = new FileSystemFake()
+
+      await fileSystem.mkdir('/abc')
+
+      await Promise.all([
+        fileSystem.writeFile('/abc/retry-abc-123.json', 'swapped span id and timestamp'),
+        fileSystem.writeFile('/abc/retry-123-abc.txt', 'invalid extension'),
+        fileSystem.writeFile('/abc/:)', 'completely wrong'),
+        fileSystem.writeFile('/abc/retry-0-a.txt', 'invalid extension'),
+        fileSystem.writeFile(`/abc/retry-${validEndTime + BigInt(2)}-ijkl.json`, JSON.stringify(payload)),
+        fileSystem.writeFile(`/abc/retry-${validEndTime + BigInt(1)}-efgh.json`, '{"invalid json :) }}}}')
+      ])
+
+      const delivery = new InMemoryDelivery()
+      const directory = new RetryQueueDirectory(fileSystem, '/abc')
+      const queue = new FileBasedRetryQueue(delivery, directory)
+
+      expect(await directory.files()).toStrictEqual([
+        `retry-${validEndTime + BigInt(2)}-ijkl.json`,
+        `retry-${validEndTime + BigInt(1)}-efgh.json`,
+        'retry-abc-123.json',
+        'retry-123-abc.txt',
+        ':)',
+        'retry-0-a.txt'
+      ])
+
+      await queue.flush()
+
+      expect(delivery.requests[0]).toStrictEqual(payload.body)
+      expect(delivery.requests).toHaveLength(1)
+
+      // all the files should have been deleted
+      expect(await directory.files()).toStrictEqual([])
+    })
+
+    it('deletes files after permanent delivery failure', async () => {
+      const validEndTime = BigInt(Date.now()) * BigInt(1_000_000)
+      const payload1 = createPayload({ spanId: 'a', endTimeUnixNano: validEndTime.toString() })
+      const payload2 = createPayload({ spanId: 'b', endTimeUnixNano: (validEndTime + BigInt(1)).toString() })
+      const payload3 = createPayload({ spanId: 'c', endTimeUnixNano: (validEndTime + BigInt(2)).toString() })
+
+      const fileSystem = new FileSystemFake()
+      const delivery = new InMemoryDelivery()
+      const directory = new RetryQueueDirectory(fileSystem, '/abc')
+      const queue = new FileBasedRetryQueue(delivery, directory)
+
+      queue.add(payload1, 1234)
+      await flushPromises()
+
+      queue.add(payload2, 1234)
+      await flushPromises()
+
+      queue.add(payload3, 1234)
+      await flushPromises()
+
+      expect(await directory.files()).toStrictEqual([
+        `retry-${validEndTime + BigInt(2)}-c.json`,
+        `retry-${validEndTime + BigInt(1)}-b.json`,
+        `retry-${validEndTime}-a.json`
+      ])
+
+      delivery.setNextResponseState('failure-discard')
+      delivery.setNextResponseState('success')
```

Comment:
I'm not quite following this test or the next with regards to the response states... we have 3 payloads to deliver, the first one fails with `failure-discard` and the next one succeeds? so why are all 3 payloads removed? I'm sure I've got the logic wrong here...

Reply:
The files are deleted when either:

1. delivery fails permanently (i.e. `failure-discard`) as delivery will never succeed so the payload can only be discarded
2. delivery succeeds as otherwise we'd send the payload again on the next `flush` and get duplicates in the dashboard